text/0000-const-generic-const-fn-bounds.md

```diff
@@ -0,0 +1,650 @@
+bounds for types substituted for `T`.
+
+## Drop
+
+A notable use case of `impl const` is defining `Drop` impls.
+Since const evaluation has no side effects, there is no simple example that
+showcases `const Drop` in any useful way. Instead we create a `Drop` impl that
+has user visible side effects:
+
+```rust
+let x = Cell::new(42);
+SomeDropType(&x);
+// x is now 41
+
+struct SomeDropType<'a>(&'a Cell<u32>);
+impl const Drop for SomeDropType {
+    fn drop(&mut self) {
+        self.0.set(self.0.get() - 1);
+    }
+}
+```
+
+You are now allowed to actually let a value of `SomeDropType` get dropped within a constant
+evaluation. This means
+
+```rust
+(SomeDropType(&Cell::new(42)), 42).1
+```
+
+is now allowed, because we can prove
+that everything from the creation of the value to the destruction is const evaluable.
+
+Note that one can implement `const Drop` for structs with fields with just a regular `Drop` impl.
+This will not allow using such
+
+```rust
+struct Foo;
+impl Drop for Foo { fn drop(&mut self) {} }
+struct Bar(Foo);
+impl const Drop for Bar { fn drop(&mut self) {} }
+// cannot call with `T == Foo`, because of missing `const Drop` impl
+fn foo<T: const Drop>(t: T) {
+    // Let t run out of scope and get dropped.
+    // Would not be ok if `T` is `Bar`,
+    // because the drop glue would drop `Bar`'s `Foo` field after the `Bar::drop` had been called.
+    // This function is therefor not accept by the compiler
+}
+```
+
+## const Drop in generic code
+
+`Drop` is special in Rust. You don't need to specify `T: Drop`, but `T::drop` will still be called
+if an object of type `T` goes out of scope. This means there's an implicit assumption, that given
+an arbitrary `T`, we might call `T::drop` if `T` has a drop impl. While we can specify
+`T: const Drop` to allow calling `T::drop` in a `const fn`, this means we can't pass e.g. `u32` for
+`T`, because `u32` has no `Drop` impl. Even types that definitely need dropping, but have no
+explicit `Drop` impl (like `struct Foo(String);`) cannot be passed if `T` requires a `Drop` bound.
+
+To summarize, there are currently three ways to interact with `Drop`:
+
+* don't mention `Drop` in the parameter bounds (or mention `?Drop`, amounting to the same thing)
+    * can pass any type that fulfills the other bounds, but may never go out of scope
+* mention `Drop` in the parameter bounds
+    * can only pass types with explicit `Drop` impls, still can't drop any values in the function
+* mention `const Drop` in the parameter bounds
+    * can only pass types with explicit `const Drop` impls (so no `u32`)
+
+The language gets a new marker trait `ConstDrop` which is automatically implemented for
+
+1. any `Copy` type
+2. any aggregate type with a `const Drop` impl consisting solely of elements of 1. 2.
+
+The body of a const function is allowed to generate drop glue for types that implement `ConstDrop`.
```

Comment:
Presumably you don't have to say `const fn foo<T: ConstDrop>(...) { ... }`? Clarify that...

ISTM that since you enforce the condition `impl const Drop for SomeType => ∀typeof(field): ConstDrop` below, then we can simply take the condition needed to satisfy `ConstDrop` and use that directly instead of having a new trait `ConstDrop`. That should scale better with other effects and restrictions, e.g. we don't need `ConstNopanicDrop`. Instead, the check becomes `any(T: Copy, T: Drop)`.

Reply:
> Presumably you don't have to say `const fn foo<T: ConstDrop>(...) { ... }`?

Yes you do. You need to give some guarantees to the body. Right now the only way to let a generic value rund out of scope in a `const fn` is by making it `Copy` (so `const fn foo<T: Copy>(...) { ... }`). Since we have no way to specify your `any` condition as a trait bound, we do need a compiler magics trait that specifies that. I mean, not that having `T: Foo | Bar` wouldn't be an intersting concept, but I want this RFC to be able to progress without inventing a whole new major language concept.

> Instead, the check becomes `any(T: Copy, T: Drop)`.

Surely you mean `any(T: Copy, T: const Drop)`?

So we'd treat `struct Foo;` without a `const Drop` impl as if it had a regular `Drop` impl, because we don't know whether the user intended to stabilize its `const Drop`ness.

For arrays and tuples we can automatically bubble up the information from the fields.

Reply:
> Yes you do. You need to give some guarantees to the body.

Alternatively, you can decide to include the drop glue or not based on whether things implement `Copy` or `const Drop` (when in a const context) based on what you know when things are monomorphic.

> Surely you mean `any(T: Copy, T: const Drop)`?

Not necessarily; in particular, how do we handle the situation with a `const fn` being called in a non-const context with a regular `Drop` impl? e.g.

```rust
struct Alpha;
impl Drop for Alpha { ... }

struct Beta;

#[derive(Copy, Clone)]
struct Gamma;

const fn drop_1<T: ConstDrop>(_: T) {}
const fn drop_2<T: Drop>(_: T) {}
const fn drop_3<T>(_: T) {} // ERROR; cannot drop T.

fn main() {
    drop_1(Alpha); // Not OK presumably? That seems like a problem...
    drop_2(Alpha); // OK because Alpha: Drop.

    drop_1(Beta); // Not OK.
    drop_2(Beta); // Not OK.

    drop_1(Gamma); // OK because Gamma: Copy.
    drop_2(Gamma); // Not OK.
}
```

> So we'd treat `struct Foo;` without a `const Drop` impl as if it had a regular `Drop` impl, because we don't know whether the user intended to stabilize its `const Drop`ness.

This means that given:
```rust
struct Foo;
struct Bar(Foo);

const fn fun<T: ConstDrop>(x: T) {}
```
you cannot write `fun(Bar(Foo))` since `Bar` is neither `Copy` nor does it implement `const Drop`. If however `Bar: Drop`, e.g. by `impl const Drop for Bar { fn drop(&mut self) {} }` then you can no longer write:
```rust
fn main() {
    let bar = Bar(Foo);
    bar.0;
|   ^^^^^ cannot move out of here
}
```

Reply:
> Alternatively, you can decide to include the drop glue or not based on whether things implement `Copy` or `const Drop` (when in a const context) based on what you know when things are monomorphic.

Hmm... but if we do this, then you might run the drop glue at runtime but not in a const context giving you different results at CTFE and runtime which we want to avoid... bummer?

Reply:
On the other hand... if we start requiring `T: ConstDrop` then this will presumably creep into the APIs of the standard library and cause breakages so it may not be feasible?

Reply:
> then you can no longer write:

Ugh! Yea that's no good.

So... we *infer* `Foo: ConstDrop` when writing `impl const Drop for Foo`, but we also allow manually implementing `ConstDrop` for types, but the same rules hold that held for `impl const Drop`?

Reply:
> if we start requiring `T: ConstDrop` then this will presumably creep into the APIs of the standard library and cause breakages so it may not be feasible?

Oh right, this doesn't do the effect systemy trickery where the bound loses its constness.

Actually. we can do this. We'll make `const ConstDrop` be implemented for the types as the rules say in this RFC, but we also implement `ConstDrop` (without the `impl const ConstDrop`, just `impl ConstDrop`) for *all* types

Reply:
> but we also implement `ConstDrop` (without the `impl const ConstDrop`, just `impl ConstDrop`) for _all_ types

Hmm; but we don't allow:
```rust
trait Foo {}
struct Bar;

impl Foo for Bar {}
impl const Foo for Bar {} // overlaps with prev impl.
```
That overlap preserves coherence tho since it's a marker trait without computational content so it probably not a problem.

So the solution of having `ConstDrop` for all types should make it so that you don't actually have to write `<T: ConstDrop>`; that's nice as a matter of ergonomics. However, in a future situation where we have `T: const Trait`:
```rust
struct Foo<T>(T);

const fn mk<T>() -> T { panic!() } // dummy...

impl<T> Foo<T> {
    const BAR: u8 = { mk::<T>(); 1 };
}
```
then you presumably have to say `impl<T: const ConstDrop> Foo<T> { ... }`.

We'll need a new name for `ConstDrop`... `T: const ConstDrop` looks super weird... =P (but we can delay finding a new name...).

...Actually, ISTM that without `T: const Trait` bounds, the need to explicitly implement `ConstDrop` or mention it as a bound doesn't exist? If so, we can leave it unstable and avoid locking us into that model?

----------------------------

As a process note, I think we'll want to be experimental with what we try here especially wrt. the standard library and trying things out... how do we experiment with this on nightly without stabilizing things inadvertently?

---------------------------

PS: I'm extra incoherent / unstructured today in what I write... sorry =D

Reply:
> So the solution of having `ConstDrop` for all types should make it so that you don't actually have to write `<T: ConstDrop>`; that's nice as a matter of ergonomics.

You still need `T: ConstDrop`, how else would you permit

```rust
fn foo<T>(t: T) -> T { t }
```

to be callable with e.g. `File`, but not have

```rust
fn bar<T>(_: T) {}
```

also take a `File`?

> `const ConstDrop`

ok, that's no good. But let's paint this bikeshed before stabilization instead of during this RFC.

> As a process note, I think we'll want to be experimental with what we try here especially wrt. the standard library and trying things out... how do we experiment with this on nightly without stabilizing things inadvertently?

Good question. As long as the feature is unstable, any const fn that uses bounds will not be min const fn, and thus not reachable from stable. So we only need to care about making types `const Trait` and then getting used in user defined `const fn`, which is also fine, because users can't specifiy trait bounds on const fn without feature gates. But considering my track record with these things, I'm gonna write tests before I write any code.

Reply:
> You still need `T: ConstDrop`, how else would you permit
> 
> ```rust
> fn foo<T>(t: T) -> T { t }
> ```
> 
> to be callable with e.g. `File`, but not have
> 
> ```rust
> fn bar<T>(_: T) {}
> ```
> 
> also take a `File`?

The following compiles fine *([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ff77ba69f4f62f27b9a1d5eeb8a33070))*:
```rust
trait Alpha { fn beta(&self) { dbg!(42); } }

impl<T> Alpha for T {}

fn gamma<T>(x: T) { x.beta(); } // Notice how we don't say `T: Alpha`.

fn main() {
    gamma(true);
}
```
Presumably the situation would be the same with `ConstDrop`. Because it is implemented for all types, you should be able to elide `<T: ConstDrop>`. When you actually are in a const context, then you must prove that you have `T: const ConstDrop` (or rather for a specific monomorphic `MyT` because we don't have `T: const ConstDrop` syntactically). At least this is how I see it working ideally, but perhaps that isn't workable?

> But let's paint this bikeshed before stabilization instead of during this RFC.

Yea, that's what I mean by "delay"... ;)

> But considering my track record with these things, I'm gonna write tests before I write any code.

Your track record is pretty good, but writing tests first sgtm!

Reply:
>  When you actually are in a const context, then you must prove that you have `T: const ConstDrop` (or rather for a specific monomorphic `MyT` because we don't have `T: const ConstDrop` syntactically). At least this is how I see it working ideally, but perhaps that isn't workable?

And the way you prove that is by adding `T: ConstDrop` in a `const fn`. We don't want to do this at monomorphization time. That would mean changing the function body will suddenly stop compiling function calls, even if the function still compiled and the signature did not change.

I mean, we are doing this somewhat for `impl Trait` and `Send`, but idk, seems dangerous to go down this road, but maybe `Drop` is special enough...

Reply:
> And the way you prove that is by adding `T: ConstDrop` in a `const fn`.

What I fear is that, even with implementing `ConstDrop` for all types, it will still creep into the generic interfaces of the standard library and thereby, generic code that uses those stdlib interfaces will suddenly need `T: ConstDrop` so that they can use the stdlib... and so you break things.

> We don't want to do this at monomorphization time.

Are you saying it may lead to post-monomorphization errors? What I'm suggesting is rather using an implied bounds mechanism... e.g. we could see this as the `T: ConstDrop`  constraint being there (on *all* type parameters...), just being implied rather than explicitly denoted.

> That would mean changing the function body will suddenly stop compiling function calls, even if the function still compiled and the signature did not change.

Do you have an example in mind of a scenario where bad things happen?

> I mean, we are doing this somewhat for `impl Trait` and `Send`, but idk, seems dangerous to go down this road, but maybe `Drop` is special enough...

Yeah it's already pretty special re. `#[may_dangle]`, being a lang item, and some other interactions previously noted... ;)

Reply:
> Do you have an example in mind of a scenario where bad things happen?

If we have a function like

```rust
const fn foo<T: Clone>(t: T) -> T {
    t
}
```

and we changed it to

```rust
const fn foo<T: Clone>(t: T) -> T {
    t.clone()
}
```

that would suddenly be a breaking change if `T` has no `const Drop` impl. While the example is very contrived, basically the moment we have any user defined bounds on `T`, things like this are totally in the realm of the possible.

Reply:
> that would suddenly be a breaking change if `T` has no `const Drop` impl.

Oh yeah.. good point. More specifically, it would be a breaking change when using a `MyT`, that doesn't satisfy `MyT: const Drop`, *in a const context.* If used in a non-const context it shouldn't be a breaking change. Bummer...

I guess it's a trade-off here... On the one hand you get to elide `T: ConstDrop` in presumably a lot of places so it makes for more ergonomic usage. It's also more likely to enable more constification of the standard library. On the other hand, semver becomes slightly harder to track.

Reply:
I think we could punt on inferring `ConstDrop`, since turning on such inference is backwards compatible. We can even not stabilize it when stabilizing the rest of this RFC.

Reply:
I think punting or not ultimately depends on whether we can constify major parts of the standard library backwards compatibly or not. I think this is one of those RFCs where we just need to do loads of experimentation with real code to see how it pans out. Armchair reasoning only gets us so far... ;)

Reply:
I wish we just had the One True Drop Trait already.  You know, the one that would take `self` by some kind of owning reference, and replace drop glue entirely rather than supplementing it, along the lines of [this old RFC](https://github.com/rust-lang/rfcs/pull/1180).  Then there would be no need for a separate `ConstDrop`, since `const OneTrueDrop` would express the needed semantics ("can run drop glue in const context").

Oh well... I know that's not happening anytime soon.